[JSC] Implement `Array.prototype.toReversed` in C++

sosukesuzuki opened the pull request:
#### 053d9a84ec27095cb583274daaf41ef796c80633
<pre>
[JSC] Implement `Array.prototype.toReversed` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=284030">https://bugs.webkit.org/show_bug.cgi?id=284030</a>

Reviewed by Yusuke Suzuki.

This patch implements `Array.prototype.toReversed` in C++ based on the
implementation of `Array.prototype.reverse` and `Array.prototype.slice`

                                            TipOfTree                  Patched

array-prototype-toReversed-string        14.6644+-0.0761     ^      7.8187+-0.3245        ^ definitely 1.8756x faster
array-prototype-toReversed-int32         11.6161+-0.0813     ^      7.5432+-0.2219        ^ definitely 1.5399x faster
array-prototype-toReversed-storage       13.7082+-0.1773     ^      9.2512+-0.5436        ^ definitely 1.4818x faster
array-prototype-toReversed-double        11.7221+-0.1975     ^      6.4701+-0.4021        ^ definitely 1.8117x faster
array-prototype-toReversed-object        15.2418+-1.2822     ^      7.6469+-0.2071        ^ definitely 1.9932x faster

* JSTests/microbenchmarks/array-prototype-toReversed-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-object.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-storage.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-string.js: Added.
(test):
* JSTests/stress/array-prototype-toReversed-contiguous.js: Added.
(sameArray):
* JSTests/stress/array-prototype-toReversed-double.js: Added.
(sameArray):
* JSTests/stress/array-prototype-toReversed-int32.js: Added.
(sameArray):
* JSTests/stress/array-prototype-toReversed-storage.js: Added.
(sameArray):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(toReversed): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::holesMustForwardToPrototype): Deleted.
(JSC::isHole): Deleted.
(JSC::containsHole): Deleted.
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastToReversed):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSArrayInlines.h:
(JSC::holesMustForwardToPrototype):
(JSC::isHole):
(JSC::containsHole):

Canonical link: <a href="https://commits.webkit.org/287431@main">https://commits.webkit.org/287431@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f800390deb1d939111554d734d59387335bb919b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79721 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84252 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30735 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7008 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/62308 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/20151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82789 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42616 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/49703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26749 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29173 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72754 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85675 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78845 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6958 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/70563 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7128 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68443 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/69804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17393 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13816 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12726 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101188 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6912 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24680 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6781 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10284 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/8584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1942,6 +1923,48 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncFill, (JSGlobalObject* globalObject, Call
+    }
+
+    for (unsigned k = 0; k < length; k++) {
+        auto fromValue = thisObject->getIndex(globalObject, length - k - 1);
```
`getIndex` can throw an error. So RETURN_IF_EXCEPTION is necessary.

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1942,6 +1923,48 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncFill, (JSGlobalObject* globalObject, Call
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (UNLIKELY(length > std::numeric_limits<uint32_t>::max())) {
+        throwRangeError(globalObject, scope, "foo"_s);
```
`"foo"_s` looks wrong.

sosukesuzuki replied:
Ahhh, sorry

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -555,6 +555,85 @@ bool JSArray::fastFill(VM& vm, unsigned startIndex, unsigned endIndex, JSValue v
+        return createWithButterfly(vm, nullptr, resultStructure, butterfly);
+    }
+    case ArrayWithArrayStorage: {
+        auto& storage = *this->butterfly()->arrayStorage();
```
We need to check sparse array map existence.

Constellation commented on Source/JavaScriptCore/runtime/JSArrayInlines.h:
```diff
@@ -335,6 +335,26 @@ ALWAYS_INLINE JSValue getProperty(JSGlobalObject* globalObject, JSObject* object
+}
+
+template<typename T>
+ALWAYS_INLINE bool containsHole(T* data, unsigned length)
```
Let's make `T*` to `const T*`.

sosukesuzuki committed:
[JSC] Implement `Array.prototype.toReversed` in C++
https://bugs.webkit.org/show_bug.cgi?id=284030

Reviewed by Yusuke Suzuki.

This patch implements `Array.prototype.toReversed` in C++ based on the
implementation of `Array.prototype.reverse` and `Array.prototype.slice`

                                            TipOfTree                  Patched

array-prototype-toReversed-string        14.6644+-0.0761     ^      7.8187+-0.3245        ^ definitely 1.8756x faster
array-prototype-toReversed-int32         11.6161+-0.0813     ^      7.5432+-0.2219        ^ definitely 1.5399x faster
array-prototype-toReversed-storage       13.7082+-0.1773     ^      9.2512+-0.5436        ^ definitely 1.4818x faster
array-prototype-toReversed-double        11.7221+-0.1975     ^      6.4701+-0.4021        ^ definitely 1.8117x faster
array-prototype-toReversed-object        15.2418+-1.2822     ^      7.6469+-0.2071        ^ definitely 1.9932x faster

* JSTests/microbenchmarks/array-prototype-toReversed-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-object.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-storage.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-toReversed-string.js: Added.
(test):
* JSTests/stress/array-prototype-toReversed-contiguous.js: Added.
(sameArray):
* JSTests/stress/array-prototype-toReversed-double.js: Added.
(sameArray):
* JSTests/stress/array-prototype-toReversed-int32.js: Added.
(sameArray):
* JSTests/stress/array-prototype-toReversed-storage.js: Added.
(sameArray):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(toReversed): Deleted.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::holesMustForwardToPrototype): Deleted.
(JSC::isHole): Deleted.
(JSC::containsHole): Deleted.
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastToReversed):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSArrayInlines.h:
(JSC::holesMustForwardToPrototype):
(JSC::isHole):
(JSC::containsHole):

Canonical link: https://commits.webkit.org/287431@main

darinadler commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1942,6 +1923,49 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncFill, (JSGlobalObject* globalObject, Call
+        return { };
+    }
+
+    if (isJSArray(thisObject)) {
+        JSArray* thisArray = jsCast<JSArray*>(thisObject);
+        auto fastResult = thisArray->fastToReversed(globalObject, length);
+        if (fastResult)
+            return JSValue::encode(fastResult);
+    }
```
I slightly prefer this coding style with fewer local variables with narrower scope:
```
if (isJSArray(thisObject)) {
    if (auto fastResult = jsCast<JSArray*>(thisObject)->fastToReversed(globalObject, length))
        return JSValue::encode(fastResult);
}
```

darinadler commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1942,6 +1923,49 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncFill, (JSGlobalObject* globalObject, Call
+    auto thisValue = callFrame->thisValue().toThis(globalObject, ECMAMode::strict());
+    RETURN_IF_EXCEPTION(scope, { });
+    if (UNLIKELY(thisValue.isUndefinedOrNull()))
+        return throwVMTypeError(globalObject, scope, "Array.prototype.fill requires that |this| not be null or undefined"_s);
```
This mentions `Array.prototype.fill`, but the function here is `reversed`.

darinadler commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1942,6 +1923,49 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncFill, (JSGlobalObject* globalObject, Call
+    RETURN_IF_EXCEPTION(scope, { });
+
+    if (UNLIKELY(length > std::numeric_limits<uint32_t>::max())) {
+        throwRangeError(globalObject, scope, "Array length must be a positive integer of safe magnitude."_s);
```
Not sure I understand why the type error message mentions the function name, but this does not. Do we have a consistent style? Is "safe magnitude" a good phrase? Is it the term used in the JavaScript specification?

sosukesuzuki replied:
This is to make error messages prior to this change consistent. The screenshot is taken in the latest Safari( this is defined at https://github.com/WebKit/WebKit/blob/a10d2e295345449f0213e2141054377d23a6c41f/Source/JavaScriptCore/runtime/ArrayConstructor.cpp#L40 )

![Screenshot 2024-12-07 at 13 39 03](https://github.com/user-attachments/assets/7ed011a8-0afe-4f6f-b544-75bb0539a9b9)

darinadler commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1942,6 +1923,49 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncFill, (JSGlobalObject* globalObject, Call
+        return { };
+    }
+
+    for (unsigned k = 0; k < length; k++) {
```
Seems unnecessary to mix `uint32_t` (used above) with `unsigned` (used here). Better to pick one.
